src/queue.jl

```diff
@@ -121,31 +121,42 @@ function get_default_queue(device::ROCDevice)
     end
 end
 
-"Kills all kernels executing on the given queue, and destroys the queue."
+"""
+    kill_queue!(queue::ROCQueue; force=false)
+
+Kills all kernels executing on the given queue, destroys the queue,
+and performs internal clean up. Returns true upon success.
+
+If `force` is `true`, this will attempt to kill a queue even if a previous
+attempt has already been made. Since queue pointers may be reused by the device,
+this may cause unpredictable behaviour by destroying an active queue.
+"""
 function kill_queue!(queue::ROCQueue; force=false)
-    #queue_ptr = @atomicswap queue.queue = Ptr{HSA.Queue}(0)
-    queue_ptr = queue.queue
     if !force && !replacefield!(queue, :active, true, false, :acquire_release, :acquire).success
-        return false # We didn't destroy the queue
+        # The queue was already marked as inactive and we are not forcing it.
+        return false
     end
     @atomic queue.active = false
 
     lock(RT_LOCK) do
+        # Clean up global queue lists
         if get(DEFAULT_QUEUES, queue.device, nothing) == queue
             delete!(DEFAULT_QUEUES, queue.device)
         end
+        delete!(QUEUES, queue.queue)
```

Comment:
We already do this in the `finalizer` call above in the `ROCQueue` constructor; if we do this in `kill_queue!`, then multiple tasks might try to delete it, and some of them will fail with an error. My intention for putting it in the finalizer is to ensure that only one task does this deletion, which only happens once there are no longer any references to that queue.

Reply:
After thinking through this, I understand now why we're getting this assertion in CI. Because we call `HSA.queue_destroy` within `kill_queue!`, the HSA runtime has the right to re-use that same address to serve future queue allocations. When we allocate queues in a tight loop, that can cause a queue to be allocated with the same address as another queue which was just destroyed (but not yet deleted from `QUEUES`).

I think we should either:
- Destroy the queue in the finalizer, but set `queue.queue = C_NULL` in `kill_queue!` to signal that the queue is effectively dead (probably not preferred because now `kill_queue!` doesn't stop a runaway kernel, which is why I wrote it)
- Ensure we always check if `QUEUES` has our queue pointer before trying to access/delete `QUEUES`, and then we can do this in `kill_queue!` again like you have done here.

Reply:
I was seeing the errors not because of a tight loop, but because during testing the "HSA Async Queue Error" was deliberately triggering a QueueError and as part of its error handler it was calling `kill_queue!`. However, it looks like the `Queue` object has been GC'ed by the time `test/hsa/queue.jl` ran (and so it's finalizer hadn't run either), and so then I was seeing the reuse of the pointer during these tests.

As a third option to your suggestions, I suggest simply moving all the finalizer code to `kill_queue()`. This is OK since this function already protects against being run multiple times by its atomic check on its `active` state. (And additionally, calling `delete()` on a dictionary does not raise any errors if the key doesn't exist).